### PR TITLE
Tatokhin/Unit tests + refactored controllers and services(Provider, InstitutionStatus, PermissionsForRole) 

### PR DESCRIPTION
**Pull request by toherberg**
-Fixed existing and created additional unit tests to examine the logic of controllers and services, uppermost, Providers, and then also for InstitutionStatus and PermissionsForRole objects in our app. To achieve this, created the system with helper methods as data generators, randomizers, etc., following the example of similar features for other objects. Also produced generic methods for easy assertions of expected and actual results in situations, where we have to examine multiple conditions (e.g. - for controllers - check response type and also value inside it, both if it is a single object, collection or even data retrieved from bad request.

-Removed user Role from parameters of IProviderService Update method. Deleted check for the role during method execution because currently, our service allows Users with the "Admin" role to update any entity from the Providers list. My changes caused that provider entities can be edited only by the user-owner.

-Altered ProviderController actions to handle exceptions thrown during service usage and generate BadRequest response with a documented specific problem. Deleted code for getting current user's role, because we don't check role anymore. So the current flow is - we need specific "ProviderEdit" permission to use controller's "Update" action and also logged user's Id must be similar to UserId of Provider entity we want to change.

-------------------

### CODE REVIEW

File: OutOfSchool/OutOfSchool.WebApi.Tests/Controllers/ProviderControllerTests.cs [hunk 1]
```diff
@@ -1,519 +1,290 @@
...
+        {
+            // Arrange
+            providerService.Setup(x => x.GetAll()).ReturnsAsync(providerDtos);
+
+
+            // Act
+            var result = (await providerController.Get().ConfigureAwait(false) as ObjectResult).Value as IEnumerable<ProviderDto>;
+
+            // Assert
+            Assert.That(providerDtos.Union(result).Count, Is.EqualTo(providerDtos.Count()));
```
Comment (h4wk13) [inline OutOfSchool/OutOfSchool.WebApi.Tests/Controllers/ProviderControllerTests.cs:111]: What does count exactly says here ? 
We should check all entries for equality.
Not sure if order is matter.

Comment (toherberg) [reply OutOfSchool/OutOfSchool.WebApi.Tests/Controllers/ProviderControllerTests.cs:111]: Union method returns IEnumerable based on two other collections members, ignoring duplicates. Due to that logic, as we have same collection length in result Value as expected collection count is, my idea was that union of two totally similar collections produces third exactly same collection)

Comment (h4wk13) [reply OutOfSchool/OutOfSchool.WebApi.Tests/Controllers/ProviderControllerTests.cs:111]: ok, does that mean that 2 collections with the same number of entries are equals ?

**Review by DmyMi — APPROVED** [review]

**Review by BHy4uk — APPROVED** [review]

File: OutOfSchool/OutOfSchool.WebApi.Tests/Controllers/PermissionsForRoleControllerTests.cs [hunk 2]
```diff
@@ -65,104 +71,138 @@ public async Task GetPermissionsForRoles_WhenEmptyCollection_ReturnsNoContentRes
...
         }
 
+
         [Test]
         [TestCase("Admin")]
         public async Task GetByRoleName_WhenRoleNameIsValid_ReturnOkResultObject(string roleName)
         {
             // Arrange
-            service.Setup(x => x.GetByRole(roleName)).ReturnsAsync(permissionsForAllRoles.SingleOrDefault(x => x.RoleName == roleName));
+            var expected = permissionsForAllRoles.Where(s => s.RoleName == roleName).Select(p => p.ToModel()).FirstOrDefault();
```
Comment (h4wk13) [inline OutOfSchool/OutOfSchool.WebApi.Tests/Controllers/PermissionsForRoleControllerTests.cs:92]: `First()` ?

File: OutOfSchool/OutOfSchool.WebApi.Tests/Controllers/PermissionsForRoleControllerTests.cs [hunk 3]
```diff
@@ -65,104 +71,138 @@ public async Task GetPermissionsForRoles_WhenEmptyCollection_ReturnsNoContentRes
...
 
             // Act
             var response = await controller.GetByRoleName(roleName).ConfigureAwait(false);
 
             // Assert
-            response.GetAssertedResponseOkAndValidValue<PermissionsForRoleDTO>();
+            response.AssertResponseOkResultAndValidateValue(expected);
         }
 
         [Test]
```
Comment (h4wk13) [inline OutOfSchool/OutOfSchool.WebApi.Tests/Controllers/PermissionsForRoleControllerTests.cs:102]: please remove

File: OutOfSchool/OutOfSchool.WebApi.Tests/Controllers/PermissionsForRoleControllerTests.cs [hunk 4]
```diff
@@ -65,104 +71,138 @@ public async Task GetPermissionsForRoles_WhenEmptyCollection_ReturnsNoContentRes
...
         [TestCase("fakeRole")]
```
Comment (h4wk13) [inline OutOfSchool/OutOfSchool.WebApi.Tests/Controllers/PermissionsForRoleControllerTests.cs:103]: please use random string instead

File: OutOfSchool/OutOfSchool.WebApi.Tests/Controllers/InstitutionStatusControllerTests.cs [hunk 5]
```diff
@@ -152,60 +162,62 @@ public async Task DeleteInstitutionStatus_WhenIdIsValid_ReturnsNoContentResult(l
...
+            return new InstitutionStatus()
             {
                 Id = 1,
-                Name = "Test",
+                Name = TestDataHelper.GetRandomWords(),
             };
         }
 
-        private IEnumerable<InstitutionStatusDTO> FakeInstitutionStatuses()
+        private IEnumerable<InstitutionStatus> FakeInstitutionStatuses()
```
Comment (h4wk13) [inline OutOfSchool/OutOfSchool.WebApi.Tests/Controllers/InstitutionStatusControllerTests.cs:203]: why it is still here ?

**Pull request merged**